### PR TITLE
changes for unified cesm/ufs/ cap

### PR DESCRIPTION
**Pull request by mvertens**
### Description of changes:
This PR effects implements the following changes:
- unified cesm/ufs cap code 
- all source code changes relative to ww3_6.07.1 in `src/source` are now implemented via #ifdef CESMCOUPLED directives
(this will make it much easier to incorporated these in upcoming share code with the ufs)

### Testing: 
The following test ERS_Ld3.TL319_t061_wt061.GMOM_JRA_WD.cheyenne_intel was run to ensure that restarts are still working. Answers will change due to the removal of hard-wired dtmax, dtcpl which are now obtained via the namelist.
Test case/suite:
Test status:  more than roundoff - not clear if its climate changing.

Fixes: Did not create an issue for the hard-wiring of dtmax.

User interface (namelist or namelist defaults) changes? New namelist wav_in namelist variables `dtcfl, dtcfli, dtmax and dtmin.`

### CODE REVIEW

File: src/cpl_nuopc/wav_comp_nuopc.F90 [hunk 1]
```diff
@@ -178,18 +134,14 @@ module wav_comp_nuopc
...
-  use wav_shr_methods       , only : chkerr, state_setscalar, state_getscalar, state_diagnose, alarmInit
-  use wav_shr_methods       , only : set_component_logging, get_component_instance, log_clock_advance
+  use NUOPC_Model           , only : NUOPC_ModelGet
+  use wav_wrapper_mod       , only : t_startf, t_stopf, t_barrierf
+  use wav_wrapper_mod       , only : shr_file_getlogunit, shr_file_setlogunit
+  use wav_kind_mod          , only : r8=>shr_kind_r8, i8=>shr_kind_i8, cl=>shr_kind_cl, cs=>shr_kind_cs
+  use wav_import_export     , only : advertise_fields, realize_fields
+  use wav_import_export     , only : state_getfldptr, state_fldchk
+  use wav_shr_mod           , only : chkerr, state_setscalar, state_getscalar, state_diagnose, alarmInit, ymd2date
+  use wav_shr_mod           , only : root_task, stdout, runtype, merge_import, dbug_flag
```
Comment (alperaltuntas) [inline src/cpl_nuopc/wav_comp_nuopc.F90:144]: Another integer `stdout` is defined below in line 157, causing duplicate definition of `stdout`.

File: src/source/w3iorsmd.F90 [hunk 2]
```diff
@@ -933,6 +903,68 @@ SUBROUTINE W3IORS ( INXOUT, NDSR, DUMFPI, INTYPE, IMOD )
...
+        character(len=*), intent(out) :: fname
+
+        ! local variables
+        integer :: yy,mm,dd,hh,mn,ss,totsec
+        logical :: exists
+        logical :: lread
+        !----------------------------------------------
+
+        ! create local lread logical for clarity
+        lread = (lwrite == .false.)
```
Comment (alperaltuntas) [inline src/source/w3iorsmd.F90]: GNU compiler is complaining about the usage of `==` here. Please change it with `.eqv` operator instead.

File: src/cpl_nuopc/wav_comp_nuopc.F90 [hunk 3]
```diff
@@ -1482,24 +1216,175 @@ subroutine ModelFinalize(gcomp, rc)
...
+    integer                        :: shrlogunit
+    logical                        :: isPresent, isSet
+    character(len=CL)              :: cvalue
+    character(len=*), parameter    :: subname = '(wav_comp_nuopc:wavinit_cesm)'
+    ! -------------------------------------------------------------------
+
+    rc = ESMF_SUCCESS
+    if (dbug_flag > 5) call ESMF_LogWrite(trim(subname)//' called', ESMF_LOGMSG_INFO)
+
+    namelist /ww3_inparm/ initfile, outfreq, dtcfl, dtcfli, dtmax, dtmin
```
Comment (alperaltuntas) [inline src/cpl_nuopc/wav_comp_nuopc.F90]: This namelist declaration should be moved above the executable statements in lines 1264-1265.

**Review by alperaltuntas — CHANGES_REQUESTED** [review]

File: src/source/w3iorsmd.F90 (hunk 2 above)
Comment (mvertens) [reply src/source/w3iorsmd.F90]: Thanks for checking this. I have replaced this if more clear if/then statements.

File: src/cpl_nuopc/wav_comp_nuopc.F90 (hunk 3 above)
Comment (mvertens) [reply src/cpl_nuopc/wav_comp_nuopc.F90]: Done.

File: src/cpl_nuopc/wav_comp_nuopc.F90 (hunk 1 above)
Comment (mvertens) [reply src/cpl_nuopc/wav_comp_nuopc.F90:144]: Thanks. I've fixed this.

File: src/source/w3iorsmd.F90 (hunk 2 above)
Comment (mnlevy1981) [reply src/source/w3iorsmd.F90]: Instead of if/then, could I suggest `lread = .not. lwrite`?

**Comment by alperaltuntas** [comment]
@mvertens Thanks for addressing the comments above! There is another subtle issue that I just noticed: When I resubmit a TL319_t061.GMOM_JRA_WD job, the long_name attribute of the HS field gets recorded as  ""Stokes drift at z=0"", while it should have been "Significant wave height". Units attribute is incorrect too. Strangely, this doesn't happen at first, but happens when I resubmit a case, and even when CONTINUE_RUN remains False. From a quick visual inspection, I am not sure what the source of this behavior is, or whether this is the only symptom. 

Here is an example test case: /glade/scratch/altuntas/SMS.TL319_t061.GMOM_JRA_WD.cheyenne_intel.C.20220106_105806_l427r7

**Comment by mvertens** [comment]
@alper - I take it you do not see this behavior in the code before this PR. Could you please verify this. I'll look into this tomorrow.

File: src/source/w3iorsmd.F90 (hunk 2 above)
Comment (alperaltuntas) [reply src/source/w3iorsmd.F90]: @mnlevy1981  Agreed! I am merging this pr as is for now, but will update this later.

**Comment by alperaltuntas** [comment]
@mvertens I just checked and encountered the same behavior in the current dev/ncar branch. So, I am merging this PR, and I'll post an issue for the faulty behavior.

**Comment by mvertens** [comment]
@alperaltuntas - thanks for checking! I'm happy it's not due to this PR.

**Pull request merged**